### PR TITLE
Use Kotlin's `Instant` and `Duration` rather than our own homebrew

### LINKED ISSUE #55: Use Kotlin `Duration` instead of custom `TimeSpan`
[Kotlin 1.3.50 has been released.](https://blog.jetbrains.com/kotlin/2019/08/kotlin-1-3-50-released/)

This includes a _"new Duration and Time Measurement API"_. This means that [our custom `TimeSpan` class](https://github.com/cph-cachet/carp.core-kotlin/blob/develop/carp.common/src/commonMain/kotlin/dk/cachet/carp/common/TimeSpan.kt) can probably be phased out and be replaced with the Kotlin API.

### LINKED ISSUE #95: Should we use the Klock library for `DateTime`?
Once more advanced `DateTime` behavior is needed, we should likely consider switching to the [korlibs/klock](https://github.com/korlibs/klock/) library.

Until then, I think it is easiest to stick to our own expect/actual `DateTime` reducing the dependencies we pull in until we use more of it (probably time-based triggers in the future might/will).

### PR DESCRIPTION
**Pull request by Whathecode**
Now that Kotlin is a bit more mature, suitable multiplatform time libraries have become available.

- Replace `DateTime` with kotlinx-datetime `Instant`. (Closes #95)
- Replace `TimeSpan` with kotlin's `Duration`. (Closes #55)

`Duration` does not yet support parsing from an ISO string, but this is targeted for Kotlin 1.5.30, at which point we can serialize using this `String` representation rather than `Long`. (#287)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -248,7 +248,7 @@ if ( patientPhoneStatus.canObtainDeviceDeployment ) // True since there are no d
 {
     val deploymentInformation: MasterDeviceDeployment =
         deploymentService.getDeviceDeploymentFor( studyDeploymentId, patientPhone.roleName )
-    val deploymentDate: DateTime = deploymentInformation.lastUpdateDate // To verify correct deployment.
+    val deploymentDate: Instant = deploymentInformation.lastUpdateDate // To verify correct deployment.
```
Comment (ltj) [inline README.md:251]: val is called "*Date", yet it is a timestamp. Would rename

File: carp.clients.core/src/commonMain/kotlin/dk/cachet/carp/clients/domain/StudyRuntimeSnapshot.kt [hunk 2]
```diff
@@ -1,18 +1,18 @@
...
 import dk.cachet.carp.deployments.application.MasterDeviceDeployment
+import kotlinx.datetime.Instant
 import kotlinx.serialization.Serializable
 
 
 @Serializable
 data class StudyRuntimeSnapshot(
     val studyDeploymentId: UUID,
-    override val creationDate: DateTime,
+    override val creationDate: Instant,
```
Comment (ltj) [inline carp.clients.core/src/commonMain/kotlin/dk/cachet/carp/clients/domain/StudyRuntimeSnapshot.kt:15]: Should be `creationTime`, `createdAt` or similar. Date implies date-only data.

File: carp.common/src/commonMain/kotlin/dk/cachet/carp/common/application/devices/DeviceRegistration.kt [hunk 3]
```diff
@@ -28,7 +29,7 @@ abstract class DeviceRegistration
      */
     abstract val deviceId: String
 
-    val registrationCreationDate: DateTime = DateTime.now()
+    val registrationCreationDate: Instant = Clock.System.now()
```
Comment (ltj) [inline carp.common/src/commonMain/kotlin/dk/cachet/carp/common/application/devices/DeviceRegistration.kt:32]: Same as prev comment. I know this was the old name as well but for v1 I believe this should be renamed to reflect its a complete timestamp

File: carp.common/src/commonMain/kotlin/dk/cachet/carp/common/domain/AggregateRoot.kt [hunk 4]
```diff
@@ -12,7 +13,7 @@ abstract class AggregateRoot<TRoot, TSnapshot : Snapshot<TRoot>, TEvent : Domain
     /**
      * The date when this object was created.
      */
-    var creationDate: DateTime = DateTime.now()
+    var creationDate: Instant = Clock.System.now()
```
Comment (ltj) [inline carp.common/src/commonMain/kotlin/dk/cachet/carp/common/domain/AggregateRoot.kt:16]: variable naming

File: carp.common/src/commonMain/kotlin/dk/cachet/carp/common/domain/Snapshot.kt [hunk 5]
```diff
@@ -12,7 +12,7 @@ import dk.cachet.carp.common.application.DateTime
 @ImplementAsDataClass
 interface Snapshot<TAggregateRoot>
 {
-    val creationDate: DateTime
+    val creationDate: Instant
```
Comment (ltj) [inline carp.common/src/commonMain/kotlin/dk/cachet/carp/common/domain/Snapshot.kt:15]: val naming

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentService.kt [hunk 6]
```diff
@@ -135,7 +135,7 @@ interface DeploymentService : ApplicationService<DeploymentService, DeploymentSe
     suspend fun deploymentSuccessful(
         studyDeploymentId: UUID,
         masterDeviceRoleName: String,
-        deviceDeploymentLastUpdateDate: DateTime
+        deviceDeploymentLastUpdateDate: Instant
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentService.kt:138]: naming

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentServiceHost.kt [hunk 7]
```diff
@@ -197,7 +197,7 @@ class DeploymentServiceHost(
     override suspend fun deploymentSuccessful(
         studyDeploymentId: UUID,
         masterDeviceRoleName: String,
-        deviceDeploymentLastUpdateDate: DateTime
+        deviceDeploymentLastUpdateDate: Instant
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/application/DeploymentServiceHost.kt:200]: naming

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/domain/StudyDeploymentSnapshot.kt [hunk 8]
```diff
@@ -14,13 +14,13 @@ import kotlinx.serialization.Serializable
 @Serializable
 data class StudyDeploymentSnapshot(
     val studyDeploymentId: UUID,
-    override val creationDate: DateTime,
+    override val creationDate: Instant,
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/domain/StudyDeploymentSnapshot.kt:17]: naming

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/domain/users/ParticipantGroupSnapshot.kt [hunk 9]
```diff
@@ -16,7 +16,7 @@ import kotlinx.serialization.Serializable
  */
 @Serializable
 data class ParticipantGroupSnapshot(
-    override val creationDate: DateTime,
+    override val creationDate: Instant,
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/domain/users/ParticipantGroupSnapshot.kt:19]: naming

File: carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/infrastructure/DeploymentServiceRequest.kt [hunk 10]
```diff
@@ -68,7 +68,7 @@ sealed class DeploymentServiceRequest
     data class DeploymentSuccessful(
         val studyDeploymentId: UUID,
         val masterDeviceRoleName: String,
-        val deviceDeploymentLastUpdateDate: DateTime
+        val deviceDeploymentLastUpdateDate: Instant
```
Comment (ltj) [inline carp.deployments.core/src/commonMain/kotlin/dk/cachet/carp/deployments/infrastructure/DeploymentServiceRequest.kt:71]: naming

File: carp.deployments.core/src/commonTest/kotlin/dk/cachet/carp/deployments/DeploymentCodeSamples.kt [hunk 11]
```diff
@@ -56,7 +56,7 @@ class DeploymentCodeSamples
         {
             val deploymentInformation: MasterDeviceDeployment =
                 deploymentService.getDeviceDeploymentFor( studyDeploymentId, patientPhone.roleName )
-            val deploymentDate: DateTime = deploymentInformation.lastUpdateDate // To verify correct deployment.
+            val deploymentDate: Instant = deploymentInformation.lastUpdateDate // To verify correct deployment.
```
Comment (ltj) [inline carp.deployments.core/src/commonTest/kotlin/dk/cachet/carp/deployments/DeploymentCodeSamples.kt:59]: naming

File: carp.protocols.core/src/commonMain/kotlin/dk/cachet/carp/protocols/application/StudyProtocolSnapshot.kt [hunk 12]
```diff
@@ -20,7 +20,7 @@ import kotlinx.serialization.Serializable
 data class StudyProtocolSnapshot(
     val id: StudyProtocolId,
     val description: String,
-    override val creationDate: DateTime,
+    override val creationDate: Instant,
```
Comment (ltj) [inline carp.protocols.core/src/commonMain/kotlin/dk/cachet/carp/protocols/application/StudyProtocolSnapshot.kt:23]: naming

File: carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/RecruitmentSnapshot.kt [hunk 13]
```diff
@@ -1,18 +1,18 @@
...
 import dk.cachet.carp.studies.application.users.Participant
+import kotlinx.datetime.Instant
 import kotlinx.serialization.Serializable
 
 
 @Serializable
 data class RecruitmentSnapshot(
     val studyId: UUID,
-    override val creationDate: DateTime,
+    override val creationDate: Instant,
```
Comment (ltj) [inline carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/users/RecruitmentSnapshot.kt:15]: naming

**Review by ltj — COMMENTED** [review]
Very nice. I do think we should consider better naming wherever the term 'date' is used, even though the referenced value is a timestamp

**Comment by Whathecode** [comment]
Started a PR to incorporate the suggested timestamp renaming: https://github.com/cph-cachet/carp.core-kotlin/pull/290

**Pull request merged**